packages/openapi-parser/src/utils/bundle.ts

```diff
@@ -0,0 +1,253 @@
+ * ```ts
+ * const result = await readFile('./schemas/user.json')
+ * if (result.ok) {
+ *   console.log(result.data) // The normalized data
+ * } else {
+ *   console.log('Failed to read file')
+ * }
+ * ```
+ */
+export async function readFile(path: string): Promise<ResolveResult> {
+  try {
+    const fileContents = await fs.readFile(path, { encoding: 'utf-8' })
+
+    return {
+      ok: true,
+      data: normalize(fileContents),
+    }
+  } catch {
+    return {
+      ok: false,
+    }
+  }
+}
+
+/**
+ * Resolves a reference by fetching its content from either a remote URL or local file.
+ * @param ref - The reference string to resolve
+ * @returns A promise that resolves to either the content or an error result
+ * @example
+ * // Resolve a remote URL
+ * await resolveRef('https://example.com/schema.json')
+ * // Resolve a local file
+ * await resolveRef('./schemas/user.json')
+ * // Invalid reference returns { ok: false }
+ * await resolveRef('#/components/schemas/User')
+ */
+async function resolveRef(ref: string): Promise<ResolveResult> {
+  if (isRemoteUrl(ref)) {
+    return fetchUrl(ref)
+  }
+
+  return readFile(ref)
+}
+
+/**
+ * Retrieves a nested value from an object using an array of property segments.
+ * @param target - The target object to traverse
+ * @param segments - Array of property names representing the path to the desired value
+ * @returns The value at the specified path, or undefined if the path doesn't exist
+ * @example
+ * const obj = { foo: { bar: { baz: 42 } } };
+ * getNestedValue(obj, ['foo', 'bar', 'baz']); // returns 42
+ */
+export function getNestedValue(target: Record<string, any>, segments: string[]) {
+  return segments.reduce<any>((acc, key) => {
+    if (acc === undefined) {
+      return undefined
+    }
+    return acc[key]
+  }, target)
+}
+
+/**
+ * Resolves a reference path by combining a base path with a relative path.
+ * Handles both remote URLs and local file paths.
+ *
+ * @param base - The base path (can be a URL or local file path)
+ * @param relativePath - The relative path to resolve against the base
+ * @returns The resolved absolute path
+ * @example
+ * // Resolve remote URL
+ * resolveReferencePath('https://example.com/api/schema.json', 'user.json')
+ * // Returns: 'https://example.com/api/user.json'
+ *
+ * // Resolve local path
+ * resolveReferencePath('/path/to/schema.json', 'user.json')
+ * // Returns: '/path/to/user.json'
+ */
+function resolveReferencePath(base: string, relativePath: string) {
+  if (isRemoteUrl(relativePath)) {
+    return relativePath
+  }
+
+  if (isRemoteUrl(base)) {
+    const url = new URL(base)
+
+    const mergedPath = path.join(path.dirname(url.pathname), relativePath)
+    return new URL(mergedPath, base).toString()
+  }
+
+  return path.join(path.dirname(base), relativePath)
+}
+
+/**
+ * Bundles an OpenAPI specification by resolving all external references.
+ * This function traverses the input object recursively and replaces any external $ref
```

Comment:
@hanspagel @DemonHa should we be bundling by dereferencing? I think it would be more performant to put the $ref object into the components section (or an extension like `x-external`) and then replace the $ref. Right now if a bunch of paths share the same external $ref (pretty likely use case for external schemas) then we will be ballooning the size of the bundled document a lot.